Check if realloc fails

We often forget to check for failure of `realloc()`. That is a minor issue, but it causes coverity issues.

Example:

```
1. openscap-1.2.16/src/SCE/sce_engine.c:482: error[memleakOnRealloc]: Common realloc mistake: 'env_values' nulled but not freed upon failure 
1. openscap-1.2.16/src/SCE/sce_engine.c:410: error[memleakOnRealloc]: Common realloc mistake: 'env_values' nulled but not freed upon failure 
1. openscap-1.2.16/src/OVAL/probes/unix/xinetd.c:1463: error[memleakOnRealloc]: Common realloc mistake: 'newstr_array' nulled but not freed upon failure 
1. openscap-1.2.16/src/OVAL/probes/unix/xinetd.c:1384: error[memleakOnRealloc]: Common realloc mistake: 'string_array' nulled but not freed upon failure 
 1. openscap-1.2.16/src/OVAL/probes/probe/main.c:110: error[memleakOnRealloc]: Common realloc mistake: '___G_no_varref_ents' nulled but not freed upon failure 
```
Please check all `realloc` occurrences.

--- a/src/OVAL/oval_probe_ext.c
+++ b/src/OVAL/oval_probe_ext.c
@@ -137,13 +137,13 @@ static int oval_pdtbl_add(oval_pdtbl_t *tbl, oval_subtype_t type, int sd, const 
 	pd->sd      = sd;
 	pd->uri     = oscap_strdup(uri);
 
-	tbl->memb = realloc(tbl->memb, sizeof(oval_pd_t *) * (++tbl->count));
-
-	if (tbl->memb == NULL) {
+	void *new_memb = realloc(tbl->memb, sizeof(oval_pd_t *) * (++tbl->count));
+	if (new_memb == NULL) {
 		free(pd->uri);
 		free(pd);
 		return -1;
 	}
+	tbl->memb = new_memb;
 
 	tbl->memb[tbl->count - 1] = pd;
 

--- a/src/OVAL/oval_probe_handler.c
+++ b/src/OVAL/oval_probe_handler.c
@@ -96,7 +96,11 @@ int oval_probe_handler_set(oval_phtbl_t *phtbl, oval_subtype_t type, oval_probe_
                 }
         }
 
-        phtbl->ph = realloc(phtbl->ph, sizeof(oval_ph_t *) * ++phtbl->sz);
+        void *new_ph = realloc(phtbl->ph, sizeof(oval_ph_t *) * ++phtbl->sz);
+        if (new_ph == NULL)
+                return -1;
+        phtbl->ph = new_ph;
+
         phrec = phtbl->ph[phtbl->sz - 1] = malloc(sizeof(oval_ph_t));
         sort  = true;
 fillrec:

--- a/src/OVAL/probes/SEAP/generic/bitmap.c
+++ b/src/OVAL/probes/SEAP/generic/bitmap.c
@@ -97,11 +97,12 @@ int bitmap_set (bitmap_t *bitmap, bitmap_bitn_t bitn)
 
         if (i > bitmap->realsize) {
                 if (i <= bitmap->size) {
-			bitmap->cells = realloc(bitmap->cells, sizeof(uint32_t) * i);
-
+                        void *new_cells = realloc(bitmap->cells, sizeof(uint32_t) * i);
+                        if (new_cells == NULL)
+                                return (-1);
+                        bitmap->cells = new_cells;
                         memset (bitmap->cells + bitmap->realsize, 0,
                                 sizeof (bitmap_cell_t) * (i - bitmap->realsize));
-
                         bitmap->realsize = i;
                 } else {
                         return (1);
@@ -138,11 +139,17 @@ int bitmap_unset (bitmap_t *bitmap, bitmap_bitn_t bitn)
                 bitmap->cells[i - 1] &= ~(1 << (bitn % BITMAP_CELLSIZE));
                 --bitmap->count;
 
+                bitmap_size_t current_realsize = bitmap->realsize;
                 if (bitmap->realsize - i == 0) {
                         while (bitmap->cells[bitmap->realsize - 1] == 0)
                                 --bitmap->realsize;
 
-			bitmap->cells = realloc(bitmap->cells, sizeof(uint32_t) * bitmap->realsize);
+                        void *new_cells = realloc(bitmap->cells, sizeof(uint32_t) * bitmap->realsize);
+                        if (new_cells == NULL && bitmap->realsize > 0) {
+                                bitmap->realsize = current_realsize;
+                                return (-1);
+                        }
+                        bitmap->cells = new_cells;
                 }
         }
 

--- a/src/OVAL/probes/SEAP/generic/spb.c
+++ b/src/OVAL/probes/SEAP/generic/spb.c
@@ -94,12 +94,18 @@ int spb_add (spb_t *spb, void *buffer, size_t buflen)
         spb_size_t gend;
 
         if (spb->btotal >= spb->balloc) {
+                uint32_t current_balloc = spb->balloc;
                 if (spb->balloc < SPB_BALLOC_HIGHTRESH)
                         spb->balloc <<= 1;
                 else
                         spb->balloc  += SPB_BALLOC_ADD;
 
-		spb->buffer = realloc(spb->buffer, sizeof(spb_item_t) * spb->balloc);
+                void *new_buffer = realloc(spb->buffer, sizeof(spb_item_t) * spb->balloc);
+                if (new_buffer == NULL) {
+                        spb->balloc = current_balloc;
+                        return (-1);
+                }
+                spb->buffer = new_buffer;
         }
 
         /* XXX: assume that btotal > 0 if btotal > balloc? */
@@ -261,12 +267,17 @@ spb_size_t spb_drop_head (spb_t *spb, spb_size_t size, spb_flags_t flags)
                         if (flags & SPB_FLAG_FREE) {
                                 register uint32_t i;
 
-                                for (i = spb->btotal; i > 0; --i)
-					free(spb->buffer[i - 1].base);
+                                for (i = spb->btotal; i > 0; --i) {
+                                        free(spb->buffer[i - 1].base);
+                                        spb->buffer[i - 1].base = NULL;
+                                }
                         }
-
-			spb->buffer = realloc(spb->buffer, sizeof(spb_item_t) * SPB_DEFAULT_BALLOC);
                         spb->btotal = 0;
+
+                        void *new_buffer = realloc(spb->buffer, sizeof(spb_item_t) * SPB_DEFAULT_BALLOC);
+                        if (new_buffer == NULL)
+                                return (0);
+                        spb->buffer = new_buffer;
                         spb->balloc = SPB_DEFAULT_BALLOC;
                 } else {
                         /* free buffer 0..(b_idx - 1) */
@@ -275,27 +286,35 @@ spb_size_t spb_drop_head (spb_t *spb, spb_size_t size, spb_flags_t flags)
                         e_sub = spb->buffer[b_idx - 1].gend + 1;
 
                         if (flags & SPB_FLAG_FREE) {
-                                for (i = b_idx - 1; i > 0; --i)
-					free(spb->buffer[i - 1].base);
+                                for (i = b_idx - 1; i > 0; --i) {
+                                        free(spb->buffer[i - 1].base);
+                                        spb->buffer[i - 1].base = NULL;
+                                }
                         }
 
                         spb->btotal -= b_idx;
                         memmove (spb->buffer, spb->buffer + b_idx, sizeof (spb_item_t) * spb->btotal);
 
                         /*
-                         * Shrink the pre-allocated memory if it's more than 2 times larger than
-                         * the space needed for the remaining buffers.
-                         */
-                        if ((spb->balloc >> 1) > spb->btotal) {
-                                spb->balloc >>= 1;
-					spb->buffer = realloc(spb->buffer, sizeof(spb_item_t) * spb->balloc);
-                        }
-
-                        /*
                          * Update the global end index member in each remaining buffer item
                          */
                         for (i = spb->btotal; i > 0; --i)
                                 spb->buffer[i - 1].gend -= e_sub;
+
+                        /*
+                         * Shrink the pre-allocated memory if it's more than 2 times larger than
+                         * the space needed for the remaining buffers.
+                         */
+                        if ((spb->balloc >> 1) > spb->btotal) {
+                                uint32_t current_balloc = spb->balloc;
+                                spb->balloc >>= 1;
+                                void *new_buffer = spb->buffer = realloc(spb->buffer, sizeof(spb_item_t) * spb->balloc);
+                                if (new_buffer == NULL && spb->balloc > 0) {
+                                        spb->balloc = current_balloc;
+                                } else {
+                                        spb->buffer = new_buffer;
+                                }
+                        }
                 }
         }
 

--- a/src/OVAL/probes/SEAP/generic/strbuf.c
+++ b/src/OVAL/probes/SEAP/generic/strbuf.c
@@ -297,7 +297,14 @@ ssize_t strbuf_write (strbuf_t *buf, int fd)
 			iot  = 0;
 		}
 
-		iov = realloc (iov, sizeof (struct iovec) * iow);
+		void *new_iov = realloc (iov, sizeof (struct iovec) * iow);
+		if (new_iov == NULL) {
+			int e = errno;
+			free(iov);
+			errno = e;
+			return -1;
+		}
+		iov = new_iov;
 		ioc = 0;
 
 		while (cur != NULL && ioc < iow) {

--- a/src/OVAL/probes/SEAP/seap-message.c
+++ b/src/OVAL/probes/SEAP/seap-message.c
@@ -119,7 +119,10 @@ int SEAP_msgattr_set (SEAP_msg_t *msg, const char *attr, SEXP_t *value)
         if (value != NULL)
                 SEXP_VALIDATE(value);
 #endif
-	msg->attrs = realloc(msg->attrs, sizeof(SEAP_attr_t) * (++msg->attrs_cnt));
+        void *new_attrs = realloc(msg->attrs, sizeof(SEAP_attr_t) * (++msg->attrs_cnt));
+        if (new_attrs == NULL)
+                return -1;
+        msg->attrs = new_attrs;
         msg->attrs[msg->attrs_cnt - 1].name  = strdup (attr);
         msg->attrs[msg->attrs_cnt - 1].value = (value != NULL ? SEXP_ref (value) : NULL);
 

--- a/src/OVAL/probes/SEAP/seap-packet.c
+++ b/src/OVAL/probes/SEAP/seap-packet.c
@@ -206,8 +206,10 @@ static int SEAP_packet_sexp2msg (SEXP_t *sexp_msg, SEAP_msg_t *seap_msg)
         _A(attr_i >= (SEXP_list_length (sexp_msg) - 4)/2);
 
         seap_msg->attrs_cnt = attr_i;
-	seap_msg->attrs = realloc(seap_msg->attrs, sizeof(SEAP_attr_t) * seap_msg->attrs_cnt);
-        seap_msg->sexp      = SEXP_list_last (sexp_msg);
+        void *new_attrs = realloc(seap_msg->attrs, sizeof(SEAP_attr_t) * seap_msg->attrs_cnt);
+        if (new_attrs != NULL || seap_msg->attrs_cnt == 0)
+                seap_msg->attrs = new_attrs;
+        seap_msg->sexp = SEXP_list_last (sexp_msg);
 
         return (0);
 }

--- a/src/OVAL/probes/SEAP/sexp-manip.c
+++ b/src/OVAL/probes/SEAP/sexp-manip.c
@@ -1399,7 +1399,10 @@ SEXP_t *SEXP_list_sort(SEXP_t *list, int(*compare)(const SEXP_t *, const SEXP_t 
          */
         list_it_count = 1;
         list_it_alloc = SEXP_LISTIT_ARRAY_INIT;
-	list_it = malloc(sizeof(SEXP_list_it) * list_it_alloc);
+        list_it = malloc(sizeof(SEXP_list_it) * list_it_alloc);
+        if (list_it == NULL) {
+                return (NULL);
+        }
 
         list_it[0].block = SEXP_LCASTP(v_dsc.mem)->b_addr;
 
@@ -1421,7 +1424,14 @@ SEXP_t *SEXP_list_sort(SEXP_t *list, int(*compare)(const SEXP_t *, const SEXP_t 
                            list_it_alloc, list_it_alloc + SEXP_LISTIT_ARRAY_INC);
 
                         list_it_alloc += SEXP_LISTIT_ARRAY_INC;
-			list_it = realloc(list_it, sizeof(SEXP_list_it) * list_it_alloc);
+                        void *new_list_it = realloc(list_it, sizeof(SEXP_list_it) * list_it_alloc);
+                        if (new_list_it == NULL) {
+                                int e = errno;
+                                free(list_it);
+                                errno = e;
+                                return (NULL);
+                        }
+                        list_it = new_list_it;
                 }
 
                 /* skip to the next block */
@@ -1530,12 +1540,18 @@ void SEXP_lstack_free (SEXP_lstack_t *stack)
 SEXP_t *SEXP_lstack_push (SEXP_lstack_t *stack, SEXP_t *s_exp)
 {
         if (stack->l_real == stack->l_size) {
+
                 if (stack->l_size < SEXP_LSTACK_GROWFAST_TRESH)
                         stack->l_size <<= 1;
                 else
                         stack->l_size += SEXP_LSTACK_GROWSLOW_DIFF;
 
-		stack->l_sref = realloc(stack->l_sref, sizeof(SEXP_t *) * stack->l_size);
+                void *new_l_sref = realloc(stack->l_sref, sizeof(SEXP_t *) * stack->l_size);
+                if (new_l_sref == NULL) {
+                        stack->l_size = stack->l_real;
+                        return NULL;
+                }
+                stack->l_sref = new_l_sref;
         }
 
         stack->l_sref[stack->l_real++] = s_exp;
@@ -1564,8 +1580,12 @@ SEXP_t *SEXP_lstack_pop (SEXP_lstack_t *stack)
         }
 
         return (ref);
-resize:
-	stack->l_sref = realloc(stack->l_sref, sizeof(SEXP_t *) * stack->l_size);
+resize: ; // Yes, this is an empty statement. Only statements can be labeled
+        void *new_l_sref = realloc(stack->l_sref, sizeof(SEXP_t *) * stack->l_size);
+        if (new_l_sref == NULL && stack->l_size > 0)
+                stack->l_size = stack->l_real;
+        else
+                stack->l_sref = new_l_sref;
         return (ref);
 }
 

--- a/src/OVAL/probes/fsdev.c
+++ b/src/OVAL/probes/fsdev.c
@@ -199,15 +199,32 @@ static fsdev_t *__fsdev_init(fsdev_t *lfs)
 			continue;
 		if (i >= lfs->cnt) {
 			lfs->cnt += DEVID_ARRAY_ADD;
-			lfs->ids = realloc(lfs->ids, sizeof(dev_t) * lfs->cnt);
+			void *new_ids = realloc(lfs->ids, sizeof(dev_t) * lfs->cnt);
+			if (new_ids == NULL) {
+				e = errno;
+				free(lfs->ids);
+				free(lfs);
+				endmntent(fp);
+				errno = e;
+				return (NULL);
+			}
+			lfs->ids = new_ids;
 		}
 		memcpy(&(lfs->ids[i++]), &st.st_dev, sizeof(dev_t));
 	}
 
 	endmntent(fp);
 
-	lfs->ids = realloc(lfs->ids, sizeof(dev_t) * i);
-	lfs->cnt = (lfs->ids == NULL ? 0 : i);
+	void *new_ids = realloc(lfs->ids, sizeof(dev_t) * i);
+	if (new_ids == NULL) {
+		e = errno;
+		free(lfs->ids);
+		free(lfs);
+		errno = e;
+		return (NULL);
+	}
+	lfs->ids = new_ids;
+	lfs->cnt = i;
 
 	return (lfs);
 }
@@ -230,7 +247,15 @@ static fsdev_t *__fsdev_init(fsdev_t *lfs)
 	}
 
 	if (i != lfs->cnt) {
-		lfs->ids = realloc(lfs->ids, sizeof(dev_t) * i);
+		void *new_ids = realloc(lfs->ids, sizeof(dev_t) * i);
+		if (new_ids == NULL) {
+			e = errno;
+			free(lfs->ids);
+			free(lfs);
+			errno = e;
+			return (NULL);
+		}
+		lfs->ids = new_ids;
 		lfs->cnt = i;
 	}
 
@@ -263,7 +288,7 @@ static fsdev_t *__fsdev_init(fsdev_t *lfs)
 	if (lfs->ids == NULL) {
 		e = errno;
 		free(lfs);
-                fclose(fp);
+		fclose(fp);
 		errno = e;
 		return (NULL);
 	}
@@ -277,7 +302,16 @@ static fsdev_t *__fsdev_init(fsdev_t *lfs)
 
 			if (i >= lfs->cnt) {
 				lfs->cnt += DEVID_ARRAY_ADD;
-				lfs->ids = realloc(lfs->ids, sizeof(dev_t) * lfs->cnt);
+				void *new_ids = realloc(lfs->ids, sizeof(dev_t) * lfs->cnt);
+				if (new_ids == NULL) {
+					e = errno;
+					free(lfs->ids);
+					free(lfs);
+					fclose(fp);
+					errno = e;
+					return (NULL);
+				}
+				lfs->ids = new_ids;
 			}
 
 			memcpy(&(lfs->ids[i++]), &st.st_dev, sizeof(dev_t));
@@ -286,8 +320,16 @@ static fsdev_t *__fsdev_init(fsdev_t *lfs)
 
 	fclose(fp);
 
-	lfs->ids = realloc(lfs->ids, sizeof(dev_t) * i);
-	lfs->cnt = (lfs->ids == NULL ? 0 : i);
+	void *new_ids = realloc(lfs->ids, sizeof(dev_t) * i);
+	if (new_ids == NULL) {
+		e = errno;
+		free(lfs->ids);
+		free(lfs);
+		errno = e;
+		return (NULL);
+	}
+	lfs->ids = new_ids;
+	lfs->cnt = i;
 
 	return (lfs);
 }

--- a/src/OVAL/probes/independent/environmentvariable58_probe.c
+++ b/src/OVAL/probes/independent/environmentvariable58_probe.c
@@ -298,7 +298,7 @@ static int read_environment(SEXP_t *pid_ent, SEXP_t *name_ent, probe_ctx *ctx)
 	}
 
 	if ((buffer = realloc(NULL, BUFFER_SIZE)) == NULL) {
-		dE("Can't allocate memory");
+		dE("Can't re-allocate memory");
 		closedir(d);
 		return PROBE_EFAULT;
 	}
@@ -344,12 +344,15 @@ static int read_environment(SEXP_t *pid_ent, SEXP_t *name_ent, probe_ctx *ctx)
 				ssize_t s;
 				if ((size_t)buffer_used >= buffer_size) {
 					buffer_size += BUFFER_SIZE;
-					buffer = realloc(buffer, buffer_size);
-					if (buffer == NULL) {
-						dE("Can't allocate memory");
-						exit(ENOMEM);
+					void *new_buffer = realloc(buffer, buffer_size);
+					if (new_buffer == NULL) {
+						dE("Can't re-allocate memory");
+						free(buffer);
+						closedir(d);
+						close(fd);
+						return PROBE_ENOMEM;
 					}
-
+					buffer = new_buffer;
 				}
 				s = read(fd, buffer + buffer_used, buffer_size - buffer_used);
 				if (s <= 0) {

--- a/src/OVAL/probes/independent/textfilecontent54_probe.c
+++ b/src/OVAL/probes/independent/textfilecontent54_probe.c
@@ -173,7 +173,13 @@ static int process_file(const char *prefix, const char *path, const char *file, 
 
 	do {
 		buf_size += buf_inc;
-		buf = realloc(buf, buf_size);
+		void *new_buf = realloc(buf, buf_size);
+		if (new_buf == NULL) {
+			dE("Can't re-allocate memory for file-processing buffer");
+			ret = PROBE_ENOMEM;
+			goto cleanup;
+		}
+		buf = new_buf;
 		ret = read(fd, buf + buf_used, buf_inc);
 		if (ret == -1) {
 			SEXP_t *msg;
@@ -188,8 +194,15 @@ static int process_file(const char *prefix, const char *path, const char *file, 
 		buf_used += ret;
 	} while (ret == buf_inc);
 
-	if (buf_used == buf_size)
-		buf = realloc(buf, ++buf_size);
+	if (buf_used == buf_size) {
+		void *new_buf = realloc(buf, ++buf_size);
+		if (new_buf == NULL) {
+			dE("Can't re-allocate memory");
+			ret = PROBE_ENOMEM;
+			goto cleanup;
+		}
+		buf = new_buf;
+	}
 	buf[buf_used++] = '\0';
 
 	do {

--- a/src/OVAL/probes/probe/icache.c
+++ b/src/OVAL/probes/probe/icache.c
@@ -129,7 +129,13 @@ static int icache_lookup(rbt_t *tree, int64_t item_id, probe_iqpair_t *pair) {
 		*/
 		dD("cache MISS");
 
-		cached->item = realloc(cached->item, sizeof(SEXP_t *) * ++cached->count);
+		void *new_item = realloc(cached->item, sizeof(SEXP_t *) * (cached->count + 1));
+		if (new_item == NULL) {
+			dE("Unable to re-allocate memory for cache");
+			return -1;
+		}
+		cached->count++;
+		cached->item = new_item;
 		cached->item[cached->count - 1] = pair->p.item;
 
 		/* Assign an unique item ID */

--- a/src/OVAL/probes/probe/ncache.c
+++ b/src/OVAL/probes/probe/ncache.c
@@ -154,15 +154,15 @@ SEXP_t *probe_ncache_add (probe_ncache_t *cache, const char *name)
         PROBE_NCACHE_WLOCK(cache, NULL);
 
         if (cache->size <= cache->real) {
+                void *new_name = realloc (cache->name, sizeof (SEXP_t *) * (cache->size + PROBE_NCACHE_ADD_SIZE));
+                if (new_name == NULL) {
+                        SEXP_free(ref);
+                        PROBE_NCACHE_WUNLOCK(cache);
+                        return NULL;
+                }
+                cache->name  = new_name;
                 cache->size += PROBE_NCACHE_ADD_SIZE;
-                cache->name  = realloc (cache->name, sizeof (SEXP_t *) * cache->size);
         }
-
-	/* TODO: check if this is really needed */
-	if (cache->name == NULL || cache->size <= cache->real) {
-		SEXP_free(ref);
-		return NULL;
-	}
 
         cache->name[cache->real] = ref;
         ++cache->real;

--- a/src/OVAL/probes/probe/probe_main.c
+++ b/src/OVAL/probes/probe/probe_main.c
@@ -119,8 +119,12 @@ static int probe_opthandler_varref(int option, int op, va_list args)
 	if (o_temp != NULL)
 		return (0);
 
-	OSCAP_GSYM(no_varref_ents) = realloc(OSCAP_GSYM(no_varref_ents),
-						   sizeof (char *) * ++OSCAP_GSYM(no_varref_ents_cnt));
+	void *new_no_varref_ents = realloc(OSCAP_GSYM(no_varref_ents),
+						   sizeof (char *) * (OSCAP_GSYM(no_varref_ents_cnt)+1));
+	if (new_no_varref_ents == NULL)
+		return -2;
+	OSCAP_GSYM(no_varref_ents_cnt)++;
+	OSCAP_GSYM(no_varref_ents) = new_no_varref_ents;
 	OSCAP_GSYM(no_varref_ents)[OSCAP_GSYM(no_varref_ents_cnt) - 1] = strdup(o_name);
 
 	qsort(OSCAP_GSYM(no_varref_ents), OSCAP_GSYM(no_varref_ents_cnt),

--- a/src/OVAL/probes/unix/fileextendedattribute_probe.c
+++ b/src/OVAL/probes/unix/fileextendedattribute_probe.c
@@ -238,20 +238,22 @@ static int file_cb(const char *prefix, const char *p, const char *f, void *ptr, 
 		xattr_count = llistxattr(st_path_with_prefix, NULL, 0);
 
 		if (xattr_count == 0) {
-			free(st_path_with_prefix);
-			free(xattr_buf);
-			return 0;
+			goto exit;
 		}
 
 		if (xattr_count < 0) {
 			dD("FAIL: llistxattr(%s, %p, %zu): errno=%u, %s", st_path_with_prefix, NULL, (size_t)0, errno, strerror(errno));
-			free(st_path_with_prefix);
-			return 0;
+			goto exit;
 		}
 
 		/* allocate space for xattr names */
 		xattr_buflen = xattr_count;
-		xattr_buf    = realloc(xattr_buf, sizeof(char) * xattr_buflen);
+		void *new_xattr_buf = realloc(xattr_buf, sizeof(char) * xattr_buflen);
+		if (new_xattr_buf == NULL) {
+			dE("Failed to re-allocate memory for xattr_buf");
+			goto exit;
+		}
+		xattr_buf = new_xattr_buf;
 
 		/* fill the buffer */
 		xattr_count = llistxattr(st_path_with_prefix, xattr_buf, xattr_buflen);
@@ -291,11 +293,18 @@ static int file_cb(const char *prefix, const char *p, const char *f, void *ptr, 
 				// Check possible buffer overflow
 				if (sizeof(char) * (xattr_vallen + 1) <= sizeof(char) * xattr_vallen) {
 					dE("Attribute is too long.");
-					abort();
+					free(xattr_val);
+					goto exit;
 				}
 
 				// Allocate buffer, '+1' is for trailing '\0'
-				xattr_val    = realloc(xattr_val, sizeof(char) * (xattr_vallen + 1));
+				void *new_xattr_val = realloc(xattr_val, sizeof(char) * (xattr_vallen + 1));
+				if (xattr_val == NULL) {
+					dE("Failed to allocate memory for xattr_val");
+					free(xattr_val);
+					goto exit;
+				}
+				xattr_val = new_xattr_val;
 
 				// we don't want to override space for '\0' by call of 'lgetxattr'
 				// we pass only 'xattr_vallen' instead of 'xattr_vallen + 1'
@@ -336,6 +345,7 @@ static int file_cb(const char *prefix, const char *p, const char *f, void *ptr, 
 		++i;
 	} while (xattr_buf + i < xattr_buf + xattr_buflen - 1);
 
+exit:
 	free(xattr_buf);
 	free(st_path_with_prefix);
 	return 0;

--- a/src/OVAL/probes/unix/linux/partition_probe.c
+++ b/src/OVAL/probes/unix/linux/partition_probe.c
@@ -139,11 +139,15 @@ static const char *correct_fstype(char *type)
 	return "";
 }
 
-static void add_mnt_opt(char ***mnt_opts, uint8_t mnt_ocnt, char *opt)
+static uint8_t add_mnt_opt(char ***mnt_opts, uint8_t mnt_ocnt, char *opt)
 {
-	*mnt_opts = realloc(*mnt_opts, sizeof(char *) * (mnt_ocnt + 1));
-	(*mnt_opts)[mnt_ocnt - 1] = opt;
-	(*mnt_opts)[mnt_ocnt] = NULL;
+	void *new_mnt_opts = realloc(*mnt_opts, sizeof(char *) * (mnt_ocnt + 2));
+	if (new_mnt_opts == NULL)
+		return mnt_ocnt;
+	*mnt_opts = new_mnt_opts;
+	(*mnt_opts)[mnt_ocnt] = opt;
+	(*mnt_opts)[mnt_ocnt + 1] = NULL;
+	return mnt_ocnt + 1;
 }
 
 #if defined(HAVE_BLKID_GET_TAG_VALUE)
@@ -185,7 +189,7 @@ static int collect_item(probe_ctx *ctx, oval_schema_version_t over, struct mnten
         tok = strtok_r(mnt_ent->mnt_opts, ",", &save);
 
         do {
-            add_mnt_opt(&mnt_opts, ++mnt_ocnt, tok);
+            mnt_ocnt = add_mnt_opt(&mnt_opts, mnt_ocnt, tok);
         } while ((tok = strtok_r(NULL, ",", &save)) != NULL);
 
         /*
@@ -194,13 +198,13 @@ static int collect_item(probe_ctx *ctx, oval_schema_version_t over, struct mnten
          * we must use flags got by statvfs().
          */
         if (stvfs.f_flag & MS_REMOUNT) {
-            add_mnt_opt(&mnt_opts, ++mnt_ocnt, "remount");
+            mnt_ocnt = add_mnt_opt(&mnt_opts, mnt_ocnt, "remount");
         }
         if (stvfs.f_flag & MS_BIND) {
-            add_mnt_opt(&mnt_opts, ++mnt_ocnt, "bind");
+            mnt_ocnt = add_mnt_opt(&mnt_opts, mnt_ocnt, "bind");
         }
         if (stvfs.f_flag & MS_MOVE) {
-            add_mnt_opt(&mnt_opts, ++mnt_ocnt, "move");
+            mnt_ocnt = add_mnt_opt(&mnt_opts, mnt_ocnt, "move");
         }
 
         dD("mnt_ocnt = %d, mnt_opts[mnt_ocnt]=%p", mnt_ocnt, mnt_opts[mnt_ocnt]);

--- a/src/OVAL/probes/unix/linux/rpminfo_probe.c
+++ b/src/OVAL/probes/unix/linux/rpminfo_probe.c
@@ -238,8 +238,12 @@ static int get_rpminfo(struct rpminfo_req *req, struct rpminfo_rep **rep, struct
                  * We can allocate all memory needed now because we know the number
                  * of results.
                  */
-                (*rep) = realloc (*rep, sizeof (struct rpminfo_rep) * ret);
-
+                void *new_rep = realloc (*rep, sizeof (struct rpminfo_rep) * ret);
+                if (new_rep == NULL) {
+                        ret = -2;
+                        goto ret;
+                }
+                (*rep) = new_rep;
                 for (i = 0; i < ret; ++i) {
                         pkgh = rpmdbNextIterator (match);
 
@@ -255,11 +259,13 @@ static int get_rpminfo(struct rpminfo_req *req, struct rpminfo_rep **rep, struct
                 ret = 0;
 
                 while ((pkgh = rpmdbNextIterator (match)) != NULL) {
-                        (*rep) = realloc (*rep, sizeof (struct rpminfo_rep) * ++ret);
-			if (*rep == NULL) {
-					return -1;
-			}
-			pkgh2rep(pkgh, (*rep) + (ret - 1), &keyid_regex);
+                        void *new_rep = realloc (*rep, sizeof (struct rpminfo_rep) * ++ret);
+                        if (new_rep == NULL) {
+                                ret = -2;
+                                goto ret;
+                        }
+                        (*rep) = new_rep;
+                        pkgh2rep(pkgh, (*rep) + (ret - 1), &keyid_regex);
                 }
         }
 

--- a/src/OVAL/probes/unix/process58_probe.c
+++ b/src/OVAL/probes/unix/process58_probe.c
@@ -322,7 +322,15 @@ static char **get_posix_capability(int pid, int max_cap_id) {
 
 				cap_id = oscap_string_to_enum(CapabilityType, cap_name);
 				if (cap_id > -1 && cap_id <= max_cap_id) {
-					ret = realloc(ret, (ret_index + 1) * sizeof(char *));
+					void *new_ret = realloc(ret, (ret_index + 1) * sizeof(char *));
+					if (new_ret == NULL) {
+						dE("Unable to re-allocate memory for ret");
+						cap_free(cap_name);
+						free(ret);
+						ret = NULL;
+						goto exit;
+					}
+					ret = new_ret;
 					ret[ret_index] = strdup(cap_name);
 					ret_index++;
 				}
@@ -330,9 +338,17 @@ static char **get_posix_capability(int pid, int max_cap_id) {
 			}
 		}
 	}
-	ret = realloc(ret, (ret_index + 1) * sizeof(char *));
+	void *new_ret = realloc(ret, (ret_index + 1) * sizeof(char *));
+	if (new_ret == NULL) {
+		dE("Unable to re-allocate memory for ret");
+		free(ret);
+		ret = NULL;
+		goto exit;
+	}
+	ret = new_ret;
 	ret[ret_index] = NULL;
 
+exit:
 	cap_free(pid_caps);
 	return ret;
 #else

--- a/src/OVAL/probes/unix/xinetd_probe.c
+++ b/src/OVAL/probes/unix/xinetd_probe.c
@@ -1148,8 +1148,14 @@ finish_section:
 			dD("adding new strans record to an exiting one: k=%s, cnt=%u+1",
 			   st_key, st->cnt);
 
-			st->srv = realloc(st->srv, sizeof (xiconf_service_t *) * ++(st->cnt));
-			st->srv[st->cnt - 1] = scur;
+			void *new_srv = realloc(st->srv, sizeof (xiconf_service_t *) * ++(st->cnt));
+			if (new_srv == NULL) {
+				dE("Failed to re-allocate memory for st->srv");
+				return (-1);
+			} else {
+				st->srv = new_srv;
+				st->srv[st->cnt - 1] = scur;
+			}
 		}
 
 		/*
@@ -1367,7 +1373,12 @@ int op_assign_strl(void *var, char *val)
 			continue;
 		}
 		dD("Adding new member to string array: %s", tok);
-		string_array = realloc(string_array, sizeof(char *) * (++string_array_size + 1));
+		void *new_string_array = realloc(string_array, sizeof(char *) * (++string_array_size + 1));
+		if (new_string_array == NULL) {
+			*aptr = string_array;
+			return -1;
+		}
+		string_array = new_string_array;
 		string_array[string_array_size-1] = strdup(tok);
 		string_array[string_array_size] = NULL;
 	}
@@ -1398,7 +1409,12 @@ int op_insert_strl(void *var, char *val)
 			continue;
 		}
 		dD("Adding new member to string array: %s", tok);
-		string_array = realloc(string_array, sizeof(char *) * (++string_array_size + 1));
+		void *new_string_array = realloc(string_array, sizeof(char *) * (++string_array_size + 1));
+		if (new_string_array == NULL) {
+			*aptr = string_array;
+			return -1;
+		}
+		string_array = new_string_array;
 		string_array[string_array_size-1] = strdup(tok);
 		string_array[string_array_size] = NULL;
 	}
@@ -1430,6 +1446,8 @@ int op_remove_strl(void *var, char *val)
 	} else {
 		// Allocate the new string array
 		newstr_array = malloc(sizeof(char *) * (string_array_size + 1));
+		if (newstr_array == NULL)
+			return -1;
 	}
 
 	// Create an array of strings to be removed from the array
@@ -1444,7 +1462,13 @@ int op_remove_strl(void *var, char *val)
 			continue;
 		}
 		dD("Adding new member to string array: %s", tok);
-		valstr_array = realloc(valstr_array, sizeof(char *) * (++valstr_array_size + 1));
+		void *new_valstr_array = realloc(valstr_array, sizeof(char *) * (++valstr_array_size + 1));
+		if (new_valstr_array == NULL) {
+			free(newstr_array);
+			free(valstr_array);
+			return -2;
+		}
+		valstr_array = new_valstr_array;
 		valstr_array[valstr_array_size-1] = tok;
 		valstr_array[valstr_array_size] = NULL;
 	}
@@ -1477,7 +1501,9 @@ int op_remove_strl(void *var, char *val)
 	newstr_array[newstr_array_size] = NULL;
 	free(string_array);
 	free(valstr_array);
-	newstr_array = realloc(newstr_array, sizeof(char*) * (newstr_array_size + 1));
+	void *new_newstr_array = realloc(newstr_array, sizeof(char *) * (newstr_array_size + 1));
+	if (new_newstr_array != NULL)
+		newstr_array = new_newstr_array;
 	*aptr = newstr_array;
 	return 0;
 }

--- a/src/SCE/sce_engine.c
+++ b/src/SCE/sce_engine.c
@@ -427,7 +427,13 @@ xccdf_test_result_type_t sce_engine_eval_rule(struct xccdf_policy *policy, const
 	{
 		struct xccdf_value_binding* binding = xccdf_value_binding_iterator_next(value_binding_it);
 
-		env_values = realloc(env_values, (env_value_count + 3) * sizeof(char*));
+		void *new_env_values = realloc(env_values, (env_value_count + 3) * sizeof(char *));
+		if (new_env_values == NULL) {
+			dE("Unable to re-allocate memory");
+			free_env_values(env_values, index_of_first_env_value_not_compiled_in, env_value_count);
+			return XCCDF_RESULT_ERROR;
+		}
+		env_values = new_env_values;
 
 		char* name = xccdf_value_binding_get_name(binding);
 		xccdf_value_type_t type = xccdf_value_binding_get_type(binding);
@@ -502,7 +508,13 @@ xccdf_test_result_type_t sce_engine_eval_rule(struct xccdf_policy *policy, const
 		env_value_count++;
 	}
 
-	env_values = realloc(env_values, (env_value_count + 1) * sizeof(char*));
+	void *new_env_values = realloc(env_values, (env_value_count + 1) * sizeof(char*));
+	if (new_env_values == NULL) {
+		dE("Unable to re-allocate memory");
+		free_env_values(env_values, index_of_first_env_value_not_compiled_in, env_value_count);
+		return XCCDF_RESULT_ERROR;
+	}
+	env_values = new_env_values;
 	env_values[env_value_count] = NULL;
 
 	// We open a pipe for communication with the forked process
@@ -510,7 +522,7 @@ xccdf_test_result_type_t sce_engine_eval_rule(struct xccdf_policy *policy, const
 	int stderr_pipefd[2];
 	if (pipe(stdout_pipefd) == -1 || pipe(stderr_pipefd) == -1)
 	{
-		perror("pipe");
+		dE("Error in pipe");
 		free_env_values(env_values, index_of_first_env_value_not_compiled_in, env_value_count);
 		return XCCDF_RESULT_ERROR;
 	}

--- a/src/XCCDF/xccdf_session.c
+++ b/src/XCCDF/xccdf_session.c
@@ -854,8 +854,12 @@ void xccdf_session_set_custom_oval_files(struct xccdf_session *session, char **o
 	if (oval_filenames == NULL)
 		return;
 
-	struct oval_content_resource **resources = malloc(sizeof(struct oval_content_resource *));
-	resources[0] = NULL;
+	size_t count;
+	for (count = 0; oval_filenames[count];)
+		count++;
+	struct oval_content_resource **resources = malloc((count + 1) * sizeof(struct oval_content_resource *));
+	if (resources == NULL)
+		return;
 
 	for (int i = 0; oval_filenames[i];) {
 		resources[i] = malloc(sizeof(struct oval_content_resource));
@@ -863,7 +867,6 @@ void xccdf_session_set_custom_oval_files(struct xccdf_session *session, char **o
 		resources[i]->source = oscap_source_new_from_file(oval_filenames[i]);
 		resources[i]->source_owned = true;
 		i++;
-		resources = realloc(resources, (i + 1) * sizeof(struct oval_content_resource *));
 		resources[i] = NULL;
 	}
 	session->oval.custom_resources = resources;
@@ -880,6 +883,7 @@ static int _xccdf_session_get_oval_from_model(struct xccdf_session *session)
 	bool fetch_option_suggested = false;
 	char *xccdf_path_cpy = NULL;
 	char *dir_path = NULL;
+	int res = 0;
 
 	_oval_content_resources_free(session->oval.resources);
 
@@ -930,7 +934,14 @@ static int _xccdf_session_get_oval_from_model(struct xccdf_session *session)
 			}
 			resources[idx]->source = source;
 			idx++;
-			resources = realloc(resources, (idx + 1) * sizeof(struct oval_content_resource *));
+			void *new_resources = realloc(resources, (idx + 1) * sizeof(struct oval_content_resource *));
+			if (new_resources == NULL) {
+				_oval_content_resources_free(resources);
+				free(tmp_path);
+				res = -1;
+				goto cleanup;
+			}
+			resources = new_resources;
 			resources[idx] = NULL;
 		}
 		else {
@@ -953,7 +964,14 @@ static int _xccdf_session_get_oval_from_model(struct xccdf_session *session)
 						resources[idx]->source = oscap_source_new_take_memory(data, data_size, printable_path);
 						resources[idx]->source_owned = true;
 						idx++;
-						resources = realloc(resources, (idx + 1) * sizeof(struct oval_content_resource *));
+						void * new_resources = realloc(resources, (idx + 1) * sizeof(struct oval_content_resource *));
+						if (new_resources == NULL) {
+							_oval_content_resources_free(resources);
+							free(tmp_path);
+							res = -1;
+							goto cleanup;
+						}
+						resources = new_resources;
 						resources[idx] = NULL;
 						free(tmp_path);
 						continue;
@@ -971,12 +989,13 @@ static int _xccdf_session_get_oval_from_model(struct xccdf_session *session)
 		}
 		free(tmp_path);
 	}
+	session->oval.resources = resources;
+cleanup:
 	free(dir_path);
 	oscap_file_entry_iterator_free(files_it);
 	oscap_file_entry_list_free(files);
 	free(xccdf_path_cpy);
-	session->oval.resources = resources;
-	return 0;
+	return res;
 }
 
 static void _xccdf_session_free_oval_agents(struct xccdf_session *session)
@@ -1057,7 +1076,12 @@ int xccdf_session_load_oval(struct xccdf_session *session)
 				session->oval.product_cpe : (char *) oscap_productname);
 
 		/* remember sessions */
-		session->oval.agents = realloc(session->oval.agents, (idx + 2) * sizeof(struct oval_agent_session *));
+		void *new_oval_agents = realloc(session->oval.agents, (idx + 2) * sizeof(struct oval_agent_session *));
+		if (new_oval_agents == NULL) {
+			oval_agent_destroy_session(tmp_sess);
+			return -1;
+		}
+		session->oval.agents = new_oval_agents;
 		session->oval.agents[idx] = tmp_sess;
 		session->oval.agents[idx+1] = NULL;
 

--- a/src/XCCDF_POLICY/xccdf_policy_remediate.c
+++ b/src/XCCDF_POLICY/xccdf_policy_remediate.c
@@ -854,6 +854,8 @@ static char *_comment_multiline_text(char *text)
 	const char *filler = "\n# ";
 	size_t buffer_size = strlen(text) + 1; // +1 for terminating '\0'
 	char *buffer = malloc(buffer_size);
+	if (buffer == NULL)
+		return NULL;
 	char *saveptr;
 	size_t filler_len = strlen(filler);
 	size_t result_len = 0;
@@ -879,14 +881,24 @@ static char *_comment_multiline_text(char *text)
 			if (!first) {
 				if (buffer_size < result_len + filler_len + 1) {
 					buffer_size += filler_len;
-					buffer = realloc(buffer, buffer_size);
+					void *new_buffer = realloc(buffer, buffer_size);
+					if (new_buffer == NULL) {
+						free(buffer);
+						return NULL;
+					}
+					buffer = new_buffer;
 				}
 				strncpy(buffer + result_len, filler, filler_len + 1);
 				result_len += filler_len;
 			}
 			if (buffer_size < result_len + token_len + 1) {
 					buffer_size += token_len;
-					buffer = realloc(buffer, buffer_size);
+					void *new_buffer = realloc(buffer, buffer_size);
+					if (new_buffer == NULL) {
+						free(buffer);
+						return NULL;
+					}
+					buffer = new_buffer;
 			}
 			/* Copy token to output buffer */
 			strncpy(buffer + result_len, token, token_len + 1);

--- a/src/common/oscap_buffer.c
+++ b/src/common/oscap_buffer.c
@@ -85,8 +85,14 @@ void oscap_buffer_append_binary_data(struct oscap_buffer *s, const char *data, c
 		 * rather than only needed capacity in order to not fragment
 		 * the memory.
 		 */
+		size_t current_capacity = s->capacity;
 		s->capacity = ((s->capacity + append_length - 1) / INITIAL_CAPACITY + 1) * INITIAL_CAPACITY;
-		s->data = realloc(s->data, s->capacity);
+		void *new_data = realloc(s->data, s->capacity);
+		if (new_data == NULL) {
+			s->capacity = current_capacity;
+			return;
+		}
+		s->data = new_data;
 	}
 
 	memcpy(s->data + s->length, data, append_length);

